### PR TITLE
Added the barriers to fail the task and workflows with large output

### PR DESCRIPTION
**Pull request by pctreddy**
Also added a bug fix for FAILED_WITH_TERMINAL_ERROR Workflow output

### CODE REVIEW

File: core/src/main/java/com/netflix/conductor/service/ExecutionService.java [hunk 1]
```diff
@@ -160,8 +160,8 @@ public void updateTask(Task task) throws Exception {
 		updateTask(new TaskResult(task));
 	}
 	
-	public void updateTask(TaskResult task) throws Exception {
-		executor.updateTask(task);
+	public void updateTask(TaskResult taskResult) throws Exception {
+		executor.updateTask(taskResult);
```
Comment (pctreddy) [inline core/src/main/java/com/netflix/conductor/service/ExecutionService.java]: There is no change here apart from the name of the variable.

File: jersey/src/main/java/com/netflix/conductor/server/resources/TaskResource.java [hunk 2]
```diff
@@ -140,11 +140,11 @@ public Task getPendingTaskForWorkflow(@PathParam("workflowId") String workflowId
 
 	@POST
 	@ApiOperation("Update a task")
-	public String updateTask(TaskResult task) throws Exception {
-		logger.debug("Update Task: {} with callback time: {}", task, task.getCallbackAfterSeconds());
-		taskService.updateTask(task);
-		logger.debug("Task: {} updated successfully with callback time: {}", task, task.getCallbackAfterSeconds());
-		return "\"" + task.getTaskId() + "\"";
+	public String updateTask(TaskResult taskResult) throws Exception {
+		logger.debug("Update Task: {} with callback time: {}", taskResult, taskResult.getCallbackAfterSeconds());
```
Comment (pctreddy) [inline jersey/src/main/java/com/netflix/conductor/server/resources/TaskResource.java]: There is no change here apart from the name of the variable.

File: server/src/main/java/com/netflix/conductor/server/ConductorServer.java [hunk 3]
```diff
@@ -51,11 +33,26 @@
 import com.netflix.dyno.connectionpool.impl.lb.HostToken;
 import com.netflix.dyno.jedis.DynoJedisClient;
 import com.sun.jersey.api.client.Client;
-
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+import org.eclipse.jetty.server.Server;
```
Comment (pctreddy) [inline server/src/main/java/com/netflix/conductor/server/ConductorServer.java]: There is no change here

File: client/src/main/java/com/netflix/conductor/client/http/TaskClient.java [hunk 4]
```diff
@@ -206,6 +211,17 @@ public Task getPendingTaskForWorkflow(String workflowId, String taskReferenceNam
      */
     public void updateTask(TaskResult taskResult) {
         Preconditions.checkNotNull(taskResult, "Task result cannot be null");
+        try {
+            //TODO introduce a config param to control the size
+            if (objectMapper.writeValueAsString(taskResult).length() > (2 * 1024 * 1024)) { //There is hard coded since there is no easy way to pass a config in here
```
Comment (cyzhao) [inline client/src/main/java/com/netflix/conductor/client/http/TaskClient.java]: Change this to 3MB?

File: client/src/main/java/com/netflix/conductor/client/http/TaskClient.java [hunk 5]
```diff
@@ -206,6 +211,17 @@ public Task getPendingTaskForWorkflow(String workflowId, String taskReferenceNam
...
+                taskResult.setReasonForIncompletion("The TaskResult payload is greater than the permissible 2MB");
```
Comment (cyzhao) [inline client/src/main/java/com/netflix/conductor/client/http/TaskClient.java]: would be useful to log the size of task result as well.

File: client/src/main/java/com/netflix/conductor/client/http/TaskClient.java [hunk 6]
```diff
@@ -206,6 +211,17 @@ public Task getPendingTaskForWorkflow(String workflowId, String taskReferenceNam
...
+                taskResult.setStatus(TaskResult.Status.FAILED_WITH_TERMINAL_ERROR);
```
Comment (cyzhao) [inline client/src/main/java/com/netflix/conductor/client/http/TaskClient.java]: Also emit a metric so we can monitor and alert.

File: client/src/main/java/com/netflix/conductor/client/http/TaskClient.java (hunk 4 above)
Comment (cyzhao) [inline client/src/main/java/com/netflix/conductor/client/http/TaskClient.java]: write to a output stream proxy to get the size for memory efficiency

File: redis-persistence/src/main/java/com/netflix/conductor/dao/dynomite/RedisExecutionDAO.java [hunk 7]
```diff
@@ -204,8 +214,17 @@ public void updateTask(Task task) {
 		}
 
 		String payload = toJson(task);
+		//The payload is verified and
+		if(payload.length() > taskPayloadThreshold) {
+			task.setOutputData(null);
+			task.setReasonForIncompletion("Payload of the task larger than the permissible 5MB");
```
Comment (cyzhao) [inline redis-persistence/src/main/java/com/netflix/conductor/dao/dynomite/RedisExecutionDAO.java]: log taskPayloadThreshold, instead of 5

Comment (cyzhao) [reply redis-persistence/src/main/java/com/netflix/conductor/dao/dynomite/RedisExecutionDAO.java]: also log the actual size and emit a metric.

File: redis-persistence/src/main/java/com/netflix/conductor/dao/dynomite/RedisExecutionDAO.java [hunk 8]
```diff
@@ -500,13 +519,16 @@ private String insertOrUpdateWorkflow(Workflow workflow, boolean update) {
 			workflow.setEndTime(System.currentTimeMillis());
 		}
 		List<Task> tasks = workflow.getTasks();
-		workflow.setTasks(new LinkedList<>()); //QQ why are the tasks set to empty list ?
+		workflow.setTasks(new LinkedList<>());
 
 		String payload = toJson(workflow);
+		if(payload.length() > workflowInputPayloadThreshold) {
+			throw new ApplicationException(Code.INVALID_INPUT, String.format("Input payload larger than the allowed threshold of: %d", workflowInputPayloadThreshold));
```
Comment (cyzhao) [inline redis-persistence/src/main/java/com/netflix/conductor/dao/dynomite/RedisExecutionDAO.java]: Log the actual size and emit a metric

Comment (pctreddy) [reply redis-persistence/src/main/java/com/netflix/conductor/dao/dynomite/RedisExecutionDAO.java]: Done

**Comment by cocodee** [comment]
Is there any document on the size limit of task output/input?

**Pull request merged**